### PR TITLE
relevancy phase 1

### PR DESCRIPTION
**Pull request by Daniel-GrunbergerCA**
Signed-off-by: Daniel-GrunbergerCA <danielg@armosec.io>

## Overview

1 - Support SBOM scanning for new images
2 - Support CVE scanning for ne workloads with old images

### CODE REVIEW

File: mainhandler/vulnscan.go [hunk 1]
```diff
@@ -283,66 +291,39 @@ func (actionHandler *ActionHandler) scanWorkload(ctx context.Context, sessionObj
...
+
+	pod := getPodByWLID(ctx, workload)
 	if pod == nil {
 		logger.L().Info(fmt.Sprintf("workload %s has no podSpec, skipping", actionHandler.wlid))
 		return nil
 	}
+	mapContainerToImageID := make(map[string]string) // map of container name to image ID. Container name is unique per pod
+
+	// look for container to imageID map in the command args. If not found, look for it in the wl. If not found, get it from the pod
+	if val, ok := actionHandler.command.Args[utils.ContainerToImageIdsArg].(map[string]string); !ok {
```
Comment (amirmalka) [inline mainhandler/vulnscan.go:303]: I think you should separate map access to type casting

File: mainhandler/vulnscan.go [hunk 2]
```diff
@@ -283,66 +291,39 @@ func (actionHandler *ActionHandler) scanWorkload(ctx context.Context, sessionObj
...
+		if len(pod.Status.ContainerStatuses) == 0 {
+			// get from wl
+			mapContainerToImageID, err = actionHandler.getContainerToImageIDsFromWorkload(workload)
+			if err != nil {
+				logger.L().Ctx(ctx).Error(fmt.Sprintf("failed to get container to image ID map for workload %s with err %v", actionHandler.wlid, err))
```
Comment (amirmalka) [inline mainhandler/vulnscan.go:308]: add a return

File: mainhandler/vulnscan.go [hunk 3]
```diff
@@ -409,16 +388,193 @@ func sendWorkloadToVulnerabilityScanner(ctx context.Context, websocketScanComman
...
+	errs := actionHandler.sendCommandForContainers(ctx, containers, mapContainerToImageID, pod, sessionObj, apis.TypeCalculateSBOM)
+
+	if errs != "" {
+		return fmt.Errorf(errs)
+	}
+	return nil
+
+}
+
+func (actionHandler *ActionHandler) sendCommandForContainers(ctx context.Context, containers []ContainerData, mapContainerToImageID map[string]string, pod *corev1.Pod, sessionObj *utils.SessionObj, command apis.NotificationPolicyType) string {
```
Comment (amirmalka) [inline mainhandler/vulnscan.go:542]: let's return []error

File: utils/utils.go [hunk 4]
```diff
@@ -49,3 +53,11 @@ func InitReporterHttpClient() httputils.IHttpClient {
 	}
 	return &http.Client{}
 }
+
+func ExtractImageID(imageID string) string {
+	return strings.TrimPrefix(imageID, dockerPullableURN)
+}
+
+func GenerateInstanceID(parentApiVersion, namespace, kind, name, resourceVersion, containerName string) string {
```
Comment (amirmalka) [inline utils/utils.go:61]: add test for this function so that if someone changes the logic it will break the test

**Review by amirmalka — CHANGES_REQUESTED** [review]

File: main.go [hunk 5]
```diff
@@ -46,7 +46,7 @@ func main() {
 
 	initHttpHandlers()
 
-	sessionObj := make(chan utils.SessionObj)
+	sessionObj := make(chan utils.SessionObj, 50)
```
Comment (amirmalka) [inline main.go:49]: why 50? and why not in a const

File: mainhandler/handlerequests.go [hunk 6]
```diff
@@ -85,25 +85,31 @@ func (mainHandler *MainHandler) HandleWatchers(ctx context.Context) {
...
 
 	// get commands for scanning new images
 	commandsList := []*apis.Command{}
 	for _, imageID := range newImageIDs {
-		for _, wlid := range watchHandler.GetWlidsForImageID(imageID) {
-			cmd := buildScanCommandForWorkload(ctx, wlid, watchHandler.GetContainerToImageIDForWlid(wlid), apis.TypeCalculateSBOM)
-			commandsList = append(commandsList, cmd)
+		wlidsForImage := watchHandler.GetWlidsForImageID(imageID)
+		if len(wlidsForImage) == 0 {
+			logger.L().Ctx(ctx).Error(fmt.Sprintf("Image %s is not used by any workload", imageID))
```
Comment (amirmalka) [inline mainhandler/handlerequests.go:105]: is this really an error?

File: mainhandler/vulnscan.go [hunk 7]
```diff
@@ -317,17 +319,14 @@ func (actionHandler *ActionHandler) scanWorkload(ctx context.Context, sessionObj
...
 	if err != nil {
 		return fmt.Errorf("failed to get workloads from k8s, wlid: %s, reason: %s", actionHandler.wlid, err.Error())
 	}
 
-	errs += actionHandler.sendCommandForContainers(ctx, containers, mapContainerToImageID, pod, sessionObj, apis.TypeScanImages)
-	if errs != "" {
-		return fmt.Errorf(errs)
-	}
-	return nil
+	err = actionHandler.sendCommandForContainers(ctx, containers, mapContainerToImageID, pod, sessionObj, apis.TypeScanImages)
```
Comment (amirmalka) [inline mainhandler/vulnscan.go:327]:
```suggestion
	return actionHandler.sendCommandForContainers(ctx, containers, mapContainerToImageID, pod, sessionObj, apis.TypeScanImages)
```

File: mainhandler/vulnscan.go [hunk 8]
```diff
@@ -530,16 +529,13 @@ func (actionHandler *ActionHandler) calculateSBOM(ctx context.Context, sessionOb
 		return err
 	}
 
-	errs := actionHandler.sendCommandForContainers(ctx, containers, mapContainerToImageID, pod, sessionObj, apis.TypeCalculateSBOM)
+	err = actionHandler.sendCommandForContainers(ctx, containers, mapContainerToImageID, pod, sessionObj, apis.TypeCalculateSBOM)
```
Comment (amirmalka) [inline mainhandler/vulnscan.go:532]:
```suggestion
	return actionHandler.sendCommandForContainers(ctx, containers, mapContainerToImageID, pod, sessionObj, apis.TypeCalculateSBOM)
```

File: watcher/watcher.go [hunk 9]
```diff
@@ -111,14 +105,14 @@ func (wh *WatchHandler) GetImageIDsForSBOMCalculation() ([]string, error) {
 }
 
 func (wh *WatchHandler) GetWlidsForImageID(imageID string) []string {
-	return wh.imagesIDToWlidsToContainerToImageIDMap[imageID]
+	return wh.imagesIDToWlidsMap[imageID]
```
Comment (amirmalka) [inline watcher/watcher.go:108]: 1. even though you are not changing it, use `imageIDsMapMutex`
2. what happens when the imageID is missing missing from the map?

**Review by amirmalka — CHANGES_REQUESTED** [review]

**Comment by amirmalka** [comment]
@Daniel-GrunbergerCA is that okay that we merge to master?

File: main.go (hunk 5 above)
Comment (Daniel-GrunbergerCA) [reply main.go:49]: I spoke with @rcohencyberarmor  and this is the number we came up with in order not to take too much memory but also not to block too much when inserting

File: mainhandler/handlerequests.go (hunk 6 above)
Comment (Daniel-GrunbergerCA) [reply mainhandler/handlerequests.go:105]: yes, I don't see how we can have an image in the map and not have its workloads

**Comment by Daniel-GrunbergerCA** [comment]
We can't merge to master, we will do it after the entire flow it's in place, and integrated

**Comment by amirmalka** [comment]
@Daniel-GrunbergerCA but the target branch of this PR is to master...

**Review by amirmalka — APPROVED** [review]

**Pull request closed without merging**